### DISCUSSION TITLE
How to reference a composite variable label?

### DISCUSSION THREAD

**Discussion by adrfantini** [typst/typst#4043] [General]
I'm trying to reference a label that is constructed from a variable. Setting the label is easy, but I'm having issues referencing it

The following code fails:
```
#let lbl = "Typst"

= #lbl is cool #label("title." + lbl)

//This is #ref("title." + lbl) // NOPE
This is #ref(label("title." + lbl)) // NOPE as well
```

How should I do this?

**Comment by laurmaedje** [comment]
The problem here is that the label is not attached the heading, but to the text _within_ the heading. The label syntax has special handling for this, but the label call can't. You can fix it by moving the `#label("title." + lbl)` to the next line.

**Reply by adrfantini** [reply]
I'm not sure I understand what you mean.

If I do what you say it still fails:
```
#let lbl = "Typst"

= #lbl is cool
#let my-label = label("title." + lbl)

This is #ref("title." + lbl) // NOPE!
// This is #ref(my-label) // NOPE again
// This is #ref(label("title." + lbl)) // NOPE as well
```

The problem is in the `ref`, not in `label`, I think? In the fact that I can't seem to `ref` a string, nor to store and reference the correct label

[Test case](https://typst.app/project/rXrTS10ADiauEv-28BlX0b)

**Reply by laurmaedje** [reply]
Due to the `let` binding, the label is now not attached to the heading again, but rather just stored in `my-label`.

Try this:
```typ
#set heading(numbering: "(1)")
#let lbl = "Typst"

= #lbl is cool
#label("title." + lbl)

This is #ref(label("title." + lbl))
```

**Reply by adrfantini** [reply]
That works! Thank you!

May I suggest that maybe that's not super intuitive? The error messages as well were not very clear in this specific case

**Reply by laurmaedje** [reply]
Yeah, it's a bit of an unfortunate situation.

**Reply by adrfantini** [reply]
Is there something I can do? Open an issue?

**Reply by laurmaedje** [reply]
You can leave your input on https://github.com/typst/typst/issues/3512. Maybe we can reopen it.
